File: Lab_compet/test15.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

const int MAX_SIZE = 20;

int dx[] = {0, 0, 1, -1}; // possible moves in x-axis
int dy[] = {1, -1, 0, 0}; // possible moves in y-axis

int n, m; // size of the map
int map[MAX_SIZE][MAX_SIZE]; // the map itself

bool visited[MAX_SIZE][MAX_SIZE]; // to mark visited rooms
int count_paths = 0; // to count the number of possible paths

void dfs(int x, int y) {
    visited[x][y] = true;

    if (map[x][y] == 2) { // found the treasure room
        count_paths++;
        visited[x][y] = false; // mark the room as unvisited
        return;
    }

    for (int i = 0; i < 4; i++) { // try all possible moves
        int nx = x + dx[i];
        int ny = y + dy[i];

        if (nx >= 0 && nx < n && ny >= 0 && ny < m && map[nx][ny] != -1 && !visited[nx][ny]) {
            dfs(nx, ny); // recursively search the next room
        }
    }

    visited[x][y] = false; // mark the room as unvisited
}

int main() {
    cin >> n >> m;

    int start_x, start_y;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < m; j++) {
            cin >> map[i][j];

            if (map[i][j] == 1) { // found the entrance
                start_x = i;
                start_y = j;
            }
        }
    }

    dfs(start_x, start_y);

    cout << count_paths << endl;

    return 0;
}

File: Lab_compet/test15.py
def count_paths(matrix, visited, row, col):
    if matrix[row][col] == 2:  # found the target room
        return 1

    count = 0
    visited[row][col] = True  # mark current room as visited

    # explore all four directions
    for dx, dy in [(0, -1), (0, 1), (-1, 0), (1, 0)]:
        next_row, next_col = row + dx, col + dy
        if (0 <= next_row < len(matrix) and
            0 <= next_col < len(matrix[0]) and
            not visited[next_row][next_col] and
            matrix[next_row][next_col] != -1):
            count += count_paths(matrix, visited, next_row, next_col)

    visited[row][col] = False  # unmark current room as visited
    return count


m, n = map(int, input().split())
matrix = [list(map(int, input().split())) for _ in range(n)]
visited = [[False] * m for _ in range(n)]

# find the entrance
entrance = None
for i in range(n):
    for j in range(m):
        if matrix[i][j] == 1:
            entrance = (i, j)
            break
    if entrance:
        break

print(count_paths(matrix, visited, entrance[0], entrance[1]))
